### PR TITLE
[TOPIC-GPIO] miscellaneous API documentation updates

### PR DESCRIPTION
**Pull request by pabigot**
Make behavior and future intent more clear.

### CODE REVIEW

**Comment by pabigot** [comment]
@dbkinder could you please look at the last commit currently 45ef15c2e7923b17efa84b27eea227edd9a5d956 and tell me if there is some way to put content at the top of the API page.  Basically I'd like a block of text at the top of the "API Reference" section, but to have it come from the header.

File: include/drivers/gpio.h
Comment (carlescufi) [inline include/drivers/gpio.h]: Do you think `atomic` reflects the fact that we (will) disable interrupts as opposed to using another synchronization mechanism? I think it does, and it is sort of implied by the fact that all functions can be called from an ISR, but just checking.

Comment (pabigot) [reply include/drivers/gpio.h]: I think `atomic` means what I said it means in #18970.  Which we should reach agreement on and document.

**Review by galak — DISMISSED** [review]
I think the commit for updating the atomicity should be pulled out into its own PR that also updates the new drivers to match.

**Comment by pabigot** [comment]
> I think the commit for updating the atomicity should be pulled out into its own PR that also updates the new drivers to match.

There's a problem with that: I'm not going to change the drivers because I don't know the underlying HALs and what atomicity promises they make, and even if I do have hardware for all the drivers it's not an effective use of my time to test my inept attempts to update them.

I really think driver changes should be the responsibility of the people who know and normally maintain the drivers, so they can do things they way they want them done, rather than having to redo what somebody else did.

File: include/drivers/gpio.h
Comment (carlescufi) [reply include/drivers/gpio.h]: Fair enough. Let's use the terms as they are in #18970 and get a PR in for that.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
> I think the commit for updating the atomicity should be pulled out into its own PR that also updates the new drivers to match.

I think we can live with this. The driver updates will likely come from @mnkp, and there's a reason we have a staging branch. If you prefer what we could do is wait until the existing (new) drivers are updated to reflect the text in this PR in `topic-gpio`  and then merge the present one.

**Comment by galak** [comment]
> I think we can live with this. The driver updates will likely come from @mnkp, and there's a reason we have a staging branch. If you prefer what we could do is wait until the existing (new) drivers are updated to reflect the text in this PR in `topic-gpio` and then merge the present one.

However we want to do it is fine, we can either hold this PR until there's a merged PR with the changes or we drop the commit for that text and have @Mnkp pick it up in a PR with the atomic changes.

**Comment by carlescufi** [comment]
> > I think we can live with this. The driver updates will likely come from @mnkp, and there's a reason we have a staging branch. If you prefer what we could do is wait until the existing (new) drivers are updated to reflect the text in this PR in `topic-gpio` and then merge the present one.
> 
> However we want to do it is fine, we can either hold this PR until there's a merged PR with the changes or we drop the commit for that text and have @mnkp pick it up in a PR with the atomic changes.

@mnkp and @pabigot  what's your point of view? We should get this done this week so we can send an email to driver maintainers as soon as possible.

**Comment by pabigot** [comment]
> > However we want to do it is fine, we can either hold this PR until there's a merged PR with the changes or we drop the commit for that text and have @mnkp pick it up in a PR with the atomic changes.

> @mnkp and @pabigot what's your point of view? We should get this done this week so we can send an email to driver maintainers as soon as possible.

@galak's options both require waiting to document the requirement until all drivers are updated to comply with it.  I see no motivation for that, and strong motivation for adding the documentation now so people who look at this in the branch are informed what's expected of them.  That full specification of API behavior was not addressed before the original merge was a process failure.

Note there are two commits that specify atomicity behavior: one generically, and one at each description of expected behavior.

If consensus is to not document atomicity behavior until it's implemented I'll remove both commits.

**Comment by galak** [comment]
> @galak's options both require waiting to document the requirement until all drivers are updated to comply with it. I see no motivation for that, and strong motivation for adding the documentation now so people who look at this in the branch are informed what's expected of them. That full specification of API behavior was not addressed before the original merge was a process failure.
> 
> Note there are two commits that specify atomicity behavior: one generically, and one at each description of expected behavior.
> 
> If consensus is to not document atomicity behavior until it's implemented I'll remove both commits.

Just to be clear I mean the 4 drivers that have been updated.  I'm assuming that should be a relatively quick thing for @mnkp to handle.

**Comment by mnkp** [comment]
> That full specification of API behavior was not addressed before the original merge was a process failure.

There is a lot of work that needs to be done before the GPIO API is complete and clean. We've started by adapting flags used in DT to match Linux bindings. That required also some related changes to the API functions. That was the only goal of the PR that got merged on the topic branch. Which I've stated and re-stated multiple times already. Having atomicity and thread safety is important but @pabigot you're just piling extra work on top of an already huge pile. I would rather focus my time and energy on the two pin testcase to ensure that the functionality that we've implemented works as advertised. Also we need to convert more drivers. We're still not 100% sure that the we don't hit a block when implementing `gpio_port_*` functions on all the different platforms.

That said I appreciate the fact @pabigot that you've provided the proposal. I'll be happy to contribute to this issue as soon as I find some time.

**Comment by pabigot** [comment]
@mnkp Please understand, I'm **not** the one asking you to make these changes.  I think they should be done by the normal maintainers as part of a pass through confirming that the implementation is what they want to deal with in the future.  I absolutely do not understand why @galak wants the implementation changes made before we document the required behavior which (I hope) we all agree must be supported.

However, I accepted 16468 with the expectation that we would get the full definition of the API in place.  Which means defining the expected behavior and interaction with respect to interrupts and context switches, even if we haven't implemented it yet.

We should not be asking people to convert drivers, then coming back two weeks later and asking them to change them because we finally got the specification complete.

@galak please reconsider your NACK of the documentation-only PR.  This is only slowing us down.

As for the two-pin test case, AFAICT nobody's looked #19249 yet.  Yes, for right now you need to hard-code specific pins.  We can fix that up later.

**Comment by mnkp** [comment]
> We should not be asking people to convert drivers, then coming back two weeks later and asking them to change them because we finally got the specification complete.

We're doing incremental changes. It's not like we will need to throw away the code we wrote and write it again. When atomicity/thread safety concept is complete I expect I will need to review the code and throw in some `irq_lock`s in a few places. I understand that you have a different perspective since you were working on an external GPIO driver where this issues have higher impact. However, the atomicity/thread safety concept has been missing for a while now.

I would rather split the whole thing in a smaller manageable (and as much as possible independent) packages. If there are things which need to be handled together then we have no choice.

Regarding atomicity I believe we should handle the thread safety at the same time. Even if only to ensure that the returned error codes are what we want them to be. Thread safety will be required for external gpio drivers.

**Comment by pabigot** [comment]
> Regarding atomicity I believe we should handle the thread safety at the same time. Even if only to ensure that the returned error codes are what we want them to be. Thread safety will be required for external gpio drivers.

It's supposed to all be covered.  See https://github.com/zephyrproject-rtos/zephyr/issues/18530#issuecomment-530472454 and the linked general issue #18970 where the meaning of the terms is being defined.

**Comment by mnkp** [comment]
> > Regarding atomicity I believe we should handle the thread safety at the same time. Even if only to ensure that the returned error codes are what we want them to be. Thread safety will be required for external gpio drivers.
> 
> It's supposed to all be covered. See [#18530 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/18530#issuecomment-530472454) and the linked general issue #18970 where the meaning of the terms is being defined.

This PR covers atomicity only. I believe we should handle thread safety at the same time since both issues are related. If we guarantee atomicity in a normal SoC GPIO the thread safety is also insured but this doesn't hold true for the external GPIOs like gpio_sx1509b where we need to communicate over I2C bus.

**Comment by pabigot** [comment]
I hadn't intended an exception from the atomicity requirement for non-SOC implementations; however the original definitions in #18970 improperly required it to encompass "non-blocking", which was a problem.

I've updated the PR to note a requirement for both thread- and interrupt-safety for all functions.  Reentrancy is not relevant for any current function.

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: The exact same text seems to have been present in my original PR before it was removed. But it is outdated. When `gpio_pin_interrupt_configure()` function is called it needs to be passed a full specification of the interrupt it needs to configure. Passing `GPIO_INT_ENABLE` only is not enough.

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: When we talk about thread/interrupt safety we say "all functions" without going into details which one. Since we want all functions to be atomic let's not mention categories, like pin, interrupt configuration but simply refer to "all functions".

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: "toggle register"?

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: We repeat ourselves. We've mentioned already that the functions are interrupt safe and then we say "All functions may be invoked from interrupt service routines and callbacks". I would remove the sentence.

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: It should be rather "must" or "should" return `-EWOULDBLOCK`. This shouldn't be optional.

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: Current implementation of callback configuration functions (which is common to all drivers) is not thread safe, neither atomic. Someone needs to implement the feature. Should be likely in the same PR.

**Review by mnkp — DISMISSED** [review]

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: I find it extremely useful when debugging flags to be able to decode what some particular hex value maps into.

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: No; "interrupt safe" and "can be invoked from ISRs and callbacks" are two different things.

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: Anybody else have an opinion?  I can do this, though it will look weird if we define precise behavior for the `_raw()` functions and don't explain how the non-raw ones accomplish their task.

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: done

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: done

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: done

File: include/drivers/gpio.h
Comment (mnkp) [reply include/drivers/gpio.h]: The interrupt flags are mapped to the enums so this part is not really useful. In any case, you are always welcome to keep and maintain your local copy.

**Comment by pabigot** [comment]
Until #18970 gets some love and everybody agrees on what terms related to thread and interrupt safety and atomicity mean we can't document what the required behavior is.

Also @mnkp has a point about thread safety of callback registration. There are specific guarantees in the existing text about adding and removing a callback from an interrupt, which is enough for me.

I've removed the objectionable commits.  If somebody else wants some of them back, please feel free to do that in another PR.

@galak @mnkp please revisit your reviews.

**Review by galak — APPROVED** [review]

**Pull request merged**
